Readonly

samtrahan opened the pull request:
Resolves issue #6 by adding read-only detection to the `workflowdb`.  This allows `rocotostat` and `rocotocheck` to function properly on read-only databases.  Commands that require write access, such as `rocotorun`, will give error messages that users find more intuitive.  

Also, this change incorporates the fix for issue #4 because that critical bug fix was required during the test process.

samtrahan committed:
Allow Rocoto to work on read-only databases.  This will disable all
functionality that requires writing, and will prevent Rocoto from
creating tables or using the lock databases.  If the user tries to
run any Rocoto command that needs write access, the command will
complain and exit.

There is a warning message printed when the database is opened in
read-only mode.

Read-only detection happens at the top of the program by examining
the db and lock db files.  If files do not exist, they they're
assumed to be writable.  If they exist and are not writable by the
current user, then Rocoto drops to read-only mode.

Read-only can be queried by db.readonly?

samtrahan committed:
Merging bug fix from issue_4 branch

christopherwharrop committed:
Update workflowengine.rb

Fix minor comment text

christopherwharrop commented:
I'm wondering why it would not be a much simpler and cleaner solution to just add a "mode" option to the dbopen and have the status commands open the database read-only and the other commands open it read/write.  This seems overly complex.

samtrahan commented:
Chris,

There are four issues to consider:

1. The rocotostat and rocotocheck try to open the lock database and obtain
locks even though they don't need them.  Somehow the "don't open lock
database" information must be passed down.

2. Someone needs to detect whether the database should be opened in
read-only mode.

3. The error messages from rocotorun, rocotoboot, etc. were confusing.  I
tried to improve them.

4. It seemed like a good idea to warn the user that the database was opened
in read-only mode.

If you have a better idea for how to address those issues, then it's fine
with me.

Sincerely,
Sam Trahan

On Mon, Jan 9, 2017 at 11:46 AM, Christopher Harrop <
notifications@github.com> wrote:

> I'm wondering why it would not be a much simpler and cleaner solution to
> just add a "mode" option to the dbopen and have the status commands open
> the database read-only and the other commands open it read/write. This
> seems overly complex.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/christopherwharrop/rocoto/pull/7#issuecomment-271336956>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ALaK5AVlZEnNBzzBHaAcfD0swqKuLQDUks5rQmRtgaJpZM4LVDWA>
> .
>

christopherwharrop commented:
Please take a look at the issue_6 branch.  I've implemented a slightly different approach which I think is simpler and less confusing.  Please let me know if it addresses all of your concerns.  In particular, please test use of rocotostat and rocotocheck on databases owned by other users with various combinations of settings of file permissions on the database.

samtrahan commented:
Chris,

It works with all of the combinations I tried.

The only oddity is this error message given when the directory is
inaccessible:

01/10/17 15:29:14 UTC :: ERROR: Can not open /u/Samuel.Trahan/atest/a.db
read-only because it does not exist

Contrary to the message, the file does exist.  However, the user could not
read or execute the "atest" directory so the user does not know if the file
exists.

Sincerely,
Sam Trahan

On Mon, Jan 9, 2017 at 6:00 PM, Christopher Harrop <notifications@github.com
> wrote:

> Please take a look at the issue_6 branch. I've implemented a slightly
> different approach which I think is simpler and less confusing. Please let
> me know if it addresses all of your concerns. In particular, please test
> use of rocotostat and rocotocheck on databases owned by other users with
> various combinations of settings of file permissions on the database.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/christopherwharrop/rocoto/pull/7#issuecomment-271435598>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ALaK5EoGZ_LroKT2fm4oaaenLa8cbrTSks5rQrwlgaJpZM4LVDWA>
> .
>

christopherwharrop commented:
Sam,

Thank you for the testing and feedback.  I should be able to fix that error message so that it correctly states what the problem is.

christopherwharrop commented:
Sam,

I updated issue_6.  Try that scenario again and let me know if it behaves the way you expect.

samtrahan commented:
Chris,

Yes.  It works now.

./bin/rocotostat  -w ~Samuel.Trahan/atest/a.xml -d
~Samuel.Trahan/atest/a.db
01/10/17 18:08:27 UTC :: ERROR: You do not have permission to read
/u/Samuel.Trahan/atest/a.db

Sincerely,
Sam Trahan

On Tue, Jan 10, 2017 at 11:26 AM, Christopher Harrop <
notifications@github.com> wrote:

> Sam,
>
> I updated issue_6. Try that scenario again and let me know if it behaves
> the way you expect.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/christopherwharrop/rocoto/pull/7#issuecomment-271622767>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ALaK5JapnOEICCBUDTYGRH7IZS5N2yPPks5rQ7ErgaJpZM4LVDWA>
> .
>

christopherwharrop commented:
Thank you for the confirmation.  The issue_6 branch will be used to address issue #6.
